tests/run-make/reproducible-build/rmake.rs

```diff
@@ -0,0 +1,222 @@
+        rustc().arg("-g").input("reproducible-build-aux.rs").run();
+        rustc()
+            .arg("-g")
+            .input("reproducible-build.rs")
+            .linker(&cwd().join(bin_name("linker")).display().to_string())
+            .run();
+        rustc()
+            .arg("-g")
+            .input("reproducible-build.rs")
+            .linker(&cwd().join(bin_name("linker")).display().to_string())
+            .run();
+        diff().actual_file("linker-arguments1").expected_file("linker-arguments2").run();
+    });
+
+    run_in_tmpdir(|| {
+        rustc().input("linker.rs").opt().run();
+        rustc().opt().input("reproducible-build-aux.rs").run();
+        rustc()
+            .opt()
+            .input("reproducible-build.rs")
+            .linker(&cwd().join(bin_name("linker")).display().to_string())
+            .run();
+        rustc()
+            .opt()
+            .input("reproducible-build.rs")
+            .linker(&cwd().join(bin_name("linker")).display().to_string())
+            .run();
+        diff().actual_file("linker-arguments1").expected_file("linker-arguments2").run();
+    });
+
+    run_in_tmpdir(|| {
+        rustc().input("reproducible-build-aux.rs").run();
+        rustc().input("reproducible-build.rs").crate_type("rlib").library_search_path("b").run();
+        rfs::copy(rust_lib_name("reproducible_build"), rust_lib_name("foo"));
+        rustc().input("reproducible-build.rs").crate_type("rlib").library_search_path("a").run();
+        assert_eq!(rfs::read(rust_lib_name("reproducible_build")), rfs::read(rust_lib_name("foo")));
+    });
+
+    run_in_tmpdir(|| {
+        rustc().input("reproducible-build-aux.rs").run();
+        rustc()
+            .input("reproducible-build.rs")
+            .crate_type("rlib")
+            .arg("--remap-path-prefix=/a=/c")
+            .run();
+        rfs::copy(rust_lib_name("reproducible_build"), rust_lib_name("foo"));
+        rustc()
+            .input("reproducible-build.rs")
+            .crate_type("rlib")
+            .arg("--remap-path-prefix=/b=/c")
+            .run();
+        assert_eq!(rfs::read(rust_lib_name("reproducible_build")), rfs::read(rust_lib_name("foo")));
+    });
+
+    run_in_tmpdir(|| {
+        rustc().input("reproducible-build-aux.rs").run();
+        rfs::create_dir("test");
+        rfs::copy("reproducible-build.rs", "test/reproducible-build.rs");
+        rustc()
+            .input("reproducible-build.rs")
+            .crate_type("bin")
+            .arg(&format!("--remap-path-prefix={}=/b", cwd().display()))
+            .run();
+        eprintln!("{:#?}", rfs::shallow_find_dir_entries(cwd()));
+        rfs::copy(bin_name("reproducible_build"), bin_name("foo"));
+        rustc()
+            .input("test/reproducible-build.rs")
+            .crate_type("bin")
+            .arg("--remap-path-prefix=/test=/b")
+            .run();
+        assert_eq!(rfs::read(bin_name("reproducible_build")), rfs::read(bin_name("foo")));
+    });
```

Comment:
Problem: for `different_source_dirs_bin` and `different_source_dirs_rlib` cases, `remap_cwd_bin` and `remap_cwd_rlib`, some of them have tricky logic:

```make
(cd $(TMPDIR)/test && $(RUSTC) reproducible-build.rs \
    --remap-path-prefix=$(TMPDIR)/test=/b \
    --crate-type bin)
```

only changes cwd for *that* line a.k.a. that single shell invocation, and the `$(RUSTC)` here sets `-L $TMPDIR` and `--out-dir $TMPDIR`, so a case like this actually needs to do something like

```rs
// different_source_dirs_bin
run_in_tmpdir(|| {
    let base_dir = cwd();
    rustc().input("reproducible-build-aux.rs").run();
    rfs::create_dir("test");
    rfs::copy("reproducible-build.rs", "test/reproducible-build.rs");
    rustc()
        .input("reproducible-build.rs")
        .crate_type("bin")
        .arg(format!("--remap-path-prefix={}=/b", base_dir.display()))
        .run();
    rfs::copy(bin_name("reproducible-build"), bin_name("foo"));

    std::env::set_current_dir("test").unwrap();
    rustc()
        .input("reproducible-build.rs")
        .crate_type("bin")
        .arg(format!("--remap-path-prefix={}=/b", base_dir.join("test").display()))
        .library_search_path(&base_dir)
        .out_dir(&base_dir)
        .run();
    std::env::set_current_dir(&base_dir).unwrap();
    assert_eq!(
        rfs::read(base_dir.join(bin_name("foo"))),
        rfs::read(base_dir.join(bin_name("reproducible-build")))
    );
});
```

Please wrap these sections in individual functions, it was very hard to map back to the original Makefile, and they are not obvious in what they're testing.